rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp

```diff
@@ -50,17 +51,23 @@ class StaticMemoryStrategy : public memory_strategy::MemoryStrategy
   StaticMemoryStrategy(ObjectPoolBounds bounds = ObjectPoolBounds())
   : bounds_(bounds)
   {
-    memory_pool_ = static_cast<void **>(malloc(bounds_.pool_size_));
-    subscription_pool_ = static_cast<void **>(malloc(bounds_.max_subscriptions_));
-    service_pool_ = static_cast<void **>(malloc(bounds_.max_services_));
-    client_pool_ = static_cast<void **>(malloc(bounds_.max_clients_));
-    executable_pool_ = static_cast<executor::AnyExecutable *>(malloc(bounds_.max_executables_));
-
-    memset(memory_pool_, 0, bounds_.pool_size_);
-    memset(subscription_pool_, 0, bounds_.max_subscriptions_);
-    memset(service_pool_, 0, bounds_.max_services_);
-    memset(client_pool_, 0, bounds_.max_clients_);
-    memset(executable_pool_, 0, bounds_.max_executables_);
+    memory_pool_ = new void *[bounds_.pool_size_];
+    subscription_pool_ = new void *[bounds_.max_subscriptions_];
+    service_pool_ = new void *[bounds_.max_services_];
+    client_pool_ = new void *[bounds_.max_clients_];
+    guard_condition_pool_ = new void *[bounds_.max_guard_conditions_];
```

Comment:
According to http://stackoverflow.com/questions/1087042/c-new-int0-will-it-allocate-memory I don't think the current approach is a good idea.

Instead I would suggest:

    if (bounds_.pool_size_) {
      memory_pool_ = new void *[bounds_.pool_size_];
    } else {
      memory_pool_ = 0;
    }

And then later for cleaning up:

    if (memory_pool_) {
      delete[] memory_pool_;
    }